### PR TITLE
Selection questions validation

### PR DESCRIPTION
**Pull request by robyoung**
See: https://www.pivotaltracker.com/story/show/100215328

Add validation for the selection questions. Validations are hard coded in the frontend. If any validations fail the form will not be submitted to the API. Once the questions have been successfully submitted the message on the dashboard will change to "You have made the declaration".

## Validations that apply
- All required fields must be filled in (some are optional, some are optional under certain conditions).
- Character limits must not be exceeded
- Formatting rules for DUNS and VAT numbers must be followed

## Related
https://github.com/alphagov/digitalmarketplace-ssp-content/pull/53

### CODE REVIEW

File: tests/app/main/helpers/test_frameworks.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    submission['SQ1-1g'] = 'invalid'
+    assert_equal(get_all_errors(content, submission), {'SQ1-1g': 'invalid_format'})
+
+
+def test_character_limit_errors():
+    cases = [
+        ("SQ1-1a", 255),
+        ("SQ1-1cii", 255),
+        ("SQ1-1d", 255),
+        ("SQ1-1i-ii", 4096),
```
Comment (quis) [inline tests/app/main/helpers/test_frameworks.py]: We should review these character limits. They were copied from Bravo Solution™ and feel quite arbitrary—they've been removed from the content already.

Comment (quis) [inline tests/app/main/helpers/test_frameworks.py]: As discussed in meatspace, character limits can be removed.

**Comment by quis** [comment]
This is kind of grim:

image

I think it would be better if it just said
> - [Question 20](#)
> - [Question 47](#)
> - [Question 53](#)

instead.

**Comment by robyoung** [comment]
Yes, good call, now looks like...
screen shot 2015-08-11 at 16 57 44

File: app/main/helpers/frameworks.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    return required_fields
+
+
+def get_all_fields(content):
+    return reduce(add, (section.get_question_ids() for section in content))
+
+
+def get_question_numbers(content):
+    numbers = {}
+    for section in content:
```
Comment (allait) [inline app/main/helpers/frameworks.py]: You could use `get_all_fields` with `enumerate` here

File: app/main/views/frameworks.py [hunk 3]
```diff
@@ -66,20 +76,26 @@ def framework_services():
 @flask_featureflags.is_active_feature('GCLOUD7_OPEN')
```
Comment (pcraig3) [inline app/main/views/frameworks.py]: Is it easier to do `methods=['GET', 'POST']` here?  
I've generally been splitting them into different routes.

Comment (robyoung) [inline app/main/views/frameworks.py]: Marginally as both render the same template. I'm not very opinionated either way but we should be consistent. Maybe something for the tech debt wall for us to decide and change wholesale?

File: app/main/helpers/frameworks.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        "SQ1-1a": 5000,
+        "SQ1-1cii": 5000,
+        "SQ1-1d": 5000,
+        "SQ1-1i-ii": 5000,
+        "SQ3-1k": 5000,
+    }
+    errors_map = {}
+    for field, limit in length_limits.items():
+        if len(answers.get(field, "")) > limit:
+            errors_map[field] = "under_character_limit"
```
Comment (pcraig3) [inline app/main/helpers/frameworks.py]: So `under_character_limit` means they've exceeded the character limit, right?

File: app/main/helpers/frameworks.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+def get_question_numbers(content):
+    numbers = {}
+    for section in content:
+        for question_id in section.get_question_ids():
+            numbers[question_id] = len(numbers) + 1
+    return numbers
+
+
+def get_fields_with_values(answers):
+    return set(key for key, value in answers.items() if not isinstance(value, six.string_types) or len(value) > 0)
```
Comment (allait) [inline app/main/helpers/frameworks.py]: What about empty lists for `checkboxes` questions?

File: app/main/helpers/frameworks.py (hunk 4 above)
Comment (robyoung) [inline app/main/helpers/frameworks.py]: Yes, this matches the error key used for the SSP.

**Comment by allait** [comment]
:+1:

**Pull request merged**